kernel/include/priority_q.h

```diff
@@ -57,13 +49,108 @@ static ALWAYS_INLINE void z_priq_mq_remove(struct _priq_mq *pq, struct k_thread
 #define _priq_wait_best		z_priq_dumb_best
 #endif
 
+#if defined(CONFIG_64BIT)
+#define NBITS 64
+#define TRAILING_ZEROS u64_count_trailing_zeros
+#else
+#define NBITS 32
+#define TRAILING_ZEROS u32_count_trailing_zeros
+#endif /* CONFIG_64BIT */
+
+/*
+ * Return value same as e.g. memcmp
+ * > 0 -> thread 1 priority  > thread 2 priority
+ * = 0 -> thread 1 priority == thread 2 priority
+ * < 0 -> thread 1 priority  < thread 2 priority
+ * Do not rely on the actual value returned aside from the above.
+ * (Again, like memcmp.)
+ */
+static ALWAYS_INLINE int32_t z_sched_prio_cmp(struct k_thread *thread_1,
+					      struct k_thread *thread_2)
+{
+	/* `prio` is <32b, so the below cannot overflow. */
+	int32_t b1 = thread_1->base.prio;
+	int32_t b2 = thread_2->base.prio;
+
+	if (b1 != b2) {
+		return b2 - b1;
+	}
+
+#ifdef CONFIG_SCHED_DEADLINE
+	/* If we assume all deadlines live within the same "half" of
+	 * the 32 bit modulus space (this is a documented API rule),
+	 * then the latest deadline in the queue minus the earliest is
+	 * guaranteed to be (2's complement) non-negative.  We can
+	 * leverage that to compare the values without having to check
+	 * the current time.
+	 */
+	uint32_t d1 = thread_1->base.prio_deadline;
+	uint32_t d2 = thread_2->base.prio_deadline;
+
+	if (d1 != d2) {
+		/* Sooner deadline means higher effective priority.
+		 * Doing the calculation with unsigned types and casting
+		 * to signed isn't perfect, but at least reduces this
+		 * from UB on overflow to impdef.
+		 */
+		return (int32_t) (d2 - d1);
+	}
+#endif /* CONFIG_SCHED_DEADLINE */
+	return 0;
+}
+
+static ALWAYS_INLINE void z_priq_dumb_add(sys_dlist_t *pq,
+					  struct k_thread *thread)
+{
+	struct k_thread *t;
+
+	SYS_DLIST_FOR_EACH_CONTAINER(pq, t, base.qnode_dlist) {
+		if (z_sched_prio_cmp(thread, t) > 0) {
+			sys_dlist_insert(&t->base.qnode_dlist,
+					 &thread->base.qnode_dlist);
+			return;
+		}
+	}
+
+	sys_dlist_append(pq, &thread->base.qnode_dlist);
+}
+
 static ALWAYS_INLINE void z_priq_dumb_remove(sys_dlist_t *pq, struct k_thread *thread)
 {
 	ARG_UNUSED(pq);
 
 	sys_dlist_remove(&thread->base.qnode_dlist);
 }
 
+static ALWAYS_INLINE void z_priq_dumb_yield(sys_dlist_t *pq)
+{
+#ifndef CONFIG_SMP
+	sys_dnode_t *n = sys_dlist_peek_next_no_check(pq, &_current->base.qnode_dlist);
+
+	sys_dlist_dequeue(&_current->base.qnode_dlist);
```

Comment:
Yes, but sys_dlist_dequeue() will not re-initialize the node. This is relevant here because re-initializing the node in this case results in unnecessary operations.